mage_integrations/mage_integrations/sources/dynamodb/__init__.py

```diff
@@ -0,0 +1,182 @@
+from botocore.config import Config
+from botocore.exceptions import ClientError
+from datetime import datetime
+from mage_integrations.sources.base import Source, main
+from mage_integrations.sources.catalog import Catalog, CatalogEntry
+from mage_integrations.sources.utils import get_standard_metadata
+from mage_integrations.sources.constants import (
+    COLUMN_TYPE_OBJECT,
+    COLUMN_TYPE_NUMBER,
+    COLUMN_TYPE_STRING,
+)
+from singer.schema import Schema
+from typing import Dict, Generator, List
+import boto3
+# TODO: refactor this common row_to_singer_record function
+# to util file to share between different sources.
+import mage_integrations.sources.mongodb.tap_mongodb.sync_strategies.common as common
+
+LAST_EVALUATED_TABLE_NAME = 'LastEvaluatedTableName'
+LAST_EVALUATED_KEY = 'LastEvaluatedKey'
+EXCLUSIVE_STAR_KEY = 'ExclusiveStartKey'
+TABLE_NAMES = 'TableNames'
+TABLE = 'Table'
+AWS_REGION = 'aws_region'
+ATTRIBUTE_NAME = 'AttributeName'
+ATTRIBUTE_TYPE = 'AttributeType'
+KEY_SCHEMA = 'KeySchema'
+ATTRIBUTE_DEFINITIONS = 'AttributeDefinitions'
+AWS_ACCESS_KEY_ID = 'aws_access_key_id'
+AWS_SECRET_ACCESS_KEY = 'aws_secret_access_key'
+ITEMS = 'Items'
+
+
+class DynamoDb(Source):
+    @property
+    def region(self) -> str:
+        return self.config.get(AWS_REGION, 'us-west-2')
+
+    def discover_streams(self) -> List[Dict]:
+        client = self.build_client()
+
+        response = client.list_tables()
+        table_list = response.get(TABLE_NAMES)
+        while response.get(LAST_EVALUATED_TABLE_NAME) is not None:
+            response = client.list_tables(
+                ExclusiveStartTableName=response.get(LAST_EVALUATED_TABLE_NAME))
+            table_list += response.get(TABLE_NAMES)
+
+        return [dict(
+            stream=stream_id,
+            tap_stream_id=stream_id,
+        ) for stream_id in table_list]
+
+    def discover_stream(self, client, table_name):
+        '''
+        Read a single table in DynamoDB.
+        '''
+        try:
+            table_info = client.describe_table(TableName=table_name).get(TABLE, {})
+        except ClientError:
+            self.logger.error(f'Access to table {table_name} was denied, skipping')
+            return None
+        key_props = [key_schema.get(ATTRIBUTE_NAME)
+                     for key_schema in table_info.get(KEY_SCHEMA, [])]
+        properties = dict()
+        for column_schema in table_info.get(ATTRIBUTE_DEFINITIONS, []):
+            column_name = column_schema.get(ATTRIBUTE_NAME)
+            column_type = column_schema.get(ATTRIBUTE_TYPE)
+            column_types = []
+            if column_type == 'S':
+                column_types.append(COLUMN_TYPE_STRING)
+            elif column_type == 'N':
+                column_types.append(COLUMN_TYPE_NUMBER)
+            elif column_type == 'B':
+                column_types.append(COLUMN_TYPE_OBJECT)
```

Comment:
are these the only data types in dynamodb?

Reply:
One note, 'B' here actually stands for Binary. I created a new type constant.

The `AttributeType ` from describe_table only have 3 types: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_AttributeDefinition.html.

Reply:
However when I look into what are the data types DynamoDB supports in general, it listed 10 types here: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html. The different between those two is weird.